design/secondary-network-ipam.md

```diff
@@ -0,0 +1,175 @@
+
+## Design Details
+
+We need to divide the subnet range for the provisioning network
+between a set of addresses we can use for DHCP and a set for static
+IPs.
+
+We need the installer to allocate IPs for the master nodes as it
+provisions them, and to record that information in the kubernetes
+database so those same IPs are not used for other hosts later.
+
+We need to store the subnet CIDR and existing allocations in the
+kubernetes database somewhere so new IPs can be allocated when hosts
+are provisioned.
+
+The [inwinstack/ipam](https://github.com/inwinstack/ipam) controller
+provides `Pool` and `IP` resources for allocating IPs from address
+ranges. We should evaluate it to see if we can use it for managing the
+IP allocations.
+
+The machine-api-provider-baremetal controller is responsible for
+making decisions about how to configure a host, so it should request
+IPs for secondary networks, assign them to the interfaces, and pass
+the relevant data using the ignition configuration data. It will need
+to create host-specific ignition configuration resources because it
+will be different for each host. It should also set the `Machine` as
+an owner of the `IP` so that the reservation is deleted when the
+`Machine` is deleted.
+
+### Work Items
+
+1. Ensure the IP ranges for secondary networks are captured by the
+   installer and saved to the kubernetes database as `Pool` resources.
+1. Ensure the installer registers the IP allocations for masters.
+1. Ensure the IPAM service is deployed along with the other metal3
+   components.
+1. Update the metal3 machine controller to allocate IPs and create
+   host-specific ignition configurations containing the IPs.
+1. Create image to hold IPAM operator.
+1. Add IPAM operator to metal3 deployment.
+
+### Dependencies
+
+* [inwinstack/ipam](https://github.com/inwinstack/ipam)
+
+### Test Plan
+
+No special requirements
+
+### Upgrade / Downgrade Strategy
+
+Add IPAM operator to deployment configuration
+
+### Version Skew Strategy
+
+N/A
+
+## Drawbacks [optional]
+
+This further complicates the configuration for the metal3 components
+by adding yet another container/Pod/Deployment.
+
+## Alternatives
+
+We could require an external DHCP and IPAM solution for the secondary
+networks, as we do for the primary network. This complicates
+deployments and requires more services running outside of the cluster
+to know about implementation details of the cluster in order to have
+the external DHCP server pass PXE requests to the dnsmasq instance
+that is part of the metal3 deployment, and which might change hosts
+and IPs if the pod is restarted.
+
+We could monitor the DHCP reservations given by dnsmasq and ensure
+they are configured to be persistent, then also use those IPs to set
+static addresses on the hosts during provisioning. This would leave a
+reservation to be cleaned up when a host is removed, which might be
+tricky for a discovered host that is never actually provisioned.
+
```

Comment:
Or we set a single static IP on the provisioning host with the pod with a lifetime.. :)  This could be made more robust than the version I currently have set up.

Reply:
Yeah, I should link over to the other proposal here.

Reply:
The downside with dhcp is if someone moves a cable, the port will still dhcp and the persistency is only as long as the dhcp server responds and has the same data set.